pennylane/ops/op_math/condition.py

```diff
@@ -27,6 +27,19 @@
 from pennylane.ops.op_math.symbolicop import SymbolicOp
 
 
+def _add_abstract_shapes(f):
+    """Add the shapes of all the returned variables before the returned variables."""
+
+    def new_f(*args, **kwargs):
+        out = f(*args, **kwargs)
+        shapes = []
+        for x in out:
+            shapes.extend(s for s in getattr(x, "shape", ()) if qml.math.is_abstract(s))
+        return *shapes, *out
+
+    return new_f

```

Comment:
Is this helper gonna only be used here? It seems that once any element in out is not abstract, the recorder `shapes` just skip it without even having a `()`. I suspect that this wrapper assumed the input `f` to be strctly returning only abstract variables, or its returned `shapes` could be misaligned with `out`.
Could we also have a small test for this method? This decorator could probably be useful on its own independent of the context here, so having its own test suite might help future dev's understanding if they are going to reuse this helper.

Reply:
Even if something similar is used for `for_loop` and `while_loop`, there's enough uncertainty around how to do this, I'd rather keep them decoupled till we have a better understanding of potential abstractions. I'll add some extra documentation to this method though.

Reply:
added some extra documentation with an example and a disclaimer that this is unstable.